spec/unit/type/exec_spec.rb

```diff
@@ -176,6 +201,32 @@ def exec_tester(command, exitstatus = 0, rest = {})
       expect(@logs).to eq([])
     end
 
+    describe "when checks stop execution when debugging" do
+      it "should log a message with the command" do
+        output = "test output\n'#{@command}' won't be executed because of failed check unless: 'true'\n"
+        test = exec_stub({:command => @command})
+        expect(test.check_all_attributes).to eq(false)
+
+        output.split("\n").each do |line|
+          log = @logs.shift
+          expect(log.level).to eq(:debug)
+          expect(log.message).to eq(line)
+        end
+      end
+
+      it "should log a message with a redacted command if it's sensitive" do
+        output = "test output\n'[command redacted]' won't be executed because of failed check unless: '[redacted]'\n"
+        test = exec_stub({:command => @command, :sensitive_parameters => [:command, :unless]})
+        expect(test.check_all_attributes).to eq(false)
+
+        output.split("\n").each do |line|
+          log = @logs.shift
+          expect(log.level).to eq(:debug)
+          expect(log.message).to eq(line)
+        end
```

Comment:
This is a bit fragile, as it assumes puppet doesn't emit any other messages while checking attributes. Consider checking that `@logs` contains at least the lines of output you're looking for like:

```ruby
output = /test output\n'[command redacted]' won't be executed because of failed check unless: '[redacted]'\n/
expect(@logs).to include(having_attributes(level: :debug, message: output))
```

This uses rspec3's composable `include` matcher and the `have_attributes` builtin matcher.

Reply:
There's also an `an_object_having_attributes` alias to the `have_attributes` matcher, which improves readability: `expect(@logs).to include(an_object_having_attributes(...))`

Reply:
Looks good...